minimint/src/lib.rs

```diff
@@ -41,240 +44,180 @@ pub mod config;
+    }
 
-        let outcome = 'inner: loop {
-            // We either want to handle the initial step or generate a new one by receiving a
-            // message from a peer
-            let Step {
-                output,
-                fault_log,
-                messages,
-            } = match initial_step.take() {
-                Some(step) => step,
-                None => {
-                    let (peer, peer_msg) = connections.receive().await;
-                    trace!(%peer, "Received message");
-                    hb.handle_message(&peer, peer_msg)
-                        .expect("Failed to process HBBFT input")
+    /// The main consensus function:
+    /// 1. Await a new proposal event or receiving a proposal from peers
+    /// 2. Send the `ConsensusProposal` to peers
+    /// 3. Run HBBFT until a `ConsensusOutcome` can be returned
+    pub async fn run_consensus_epoch(
+        &mut self,
+        proposal: impl Future<Output = ConsensusProposal>,
+        rng: &mut (impl RngCore + CryptoRng + Clone + 'static),
+    ) -> Vec<ConsensusOutcome> {
+        // process messages until new epoch or we have a proposal
+        let mut outcomes: Vec<ConsensusOutcome> = loop {
+            match self.await_proposal_or_peer_message().await {
+                Some(msg) if self.hbbft.next_epoch() == msg.1.epoch() => {
+                    break self.handle_message(msg).await
                 }
+                Some(msg) => self.handle_message(msg).await,
+                None => break vec![],
             };
+        };
 
-            for msg in messages {
-                trace!(target = ?msg.target, "sending message");
-                connections.send(msg.target, msg.message).await;
-            }
+        let proposal = proposal.await;
+        for peer in proposal.drop_peers.iter() {
+            self.connections.ban_peer(*peer).await;
+        }
 
-            if !fault_log.is_empty() {
-                warn!(?fault_log);
-            }
+        let step = self
+            .hbbft
+            .propose(&proposal.items, rng)
+            .expect("HBBFT propose failed");
 
-            if !output.is_empty() {
-                trace!("Processed step had an output, handing it off");
-                break 'inner output;
-            }
-        };
+        for msg in step.messages {
+            self.connections.send(msg.target, msg.message).await;
+        }
+
+        while outcomes.is_empty() {
+            let msg = self.connections.receive().await;
+            outcomes = self.handle_message(msg).await;
+        }
+        outcomes
+    }
+
+    async fn await_proposal_or_peer_message(&mut self) -> Option<PeerMessage> {
+        let awaits: Vec<Pin<Box<dyn Future<Output = Option<PeerMessage>>>>> = vec![
+            Box::pin(self.consensus.transaction_notify.notified().map(|_| None)),
+            Box::pin(self.consensus.await_consensus_proposal().map(|_| None)),
+            Box::pin(self.connections.receive().map(Some)),
```

Comment:
Can we do better than creating three heap allocations per received message here? I think I did the same in other places, just wondering if there might be an easy way to pin on the stack here.